torchaudio/prototype/pipelines/_vggish/_vggish_impl.py

```diff
@@ -0,0 +1,262 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import math
+import torch
+import torchaudio
+
+
+_MEL_BREAK_FREQUENCY_HERTZ = 700.0
+_MEL_HIGH_FREQUENCY_Q = 1127.0
+
+
+_SAMPLE_RATE = 16000
+_STFT_WINDOW_LENGTH_SECONDS = 0.025
+_STFT_HOP_LENGTH_SECONDS = 0.010
+_MEL_MIN_HZ = 125
+_MEL_MAX_HZ = 7500
+_NUM_BANDS = 64
+_LOG_OFFSET = 0.01
+_EXAMPLE_WINDOW_SECONDS = 0.96  # Each example contains 96 10ms frames
+_EXAMPLE_HOP_SECONDS = 0.96  # with zero overlap.
+
+
+def _build_features_network():
+    layers = []
+
+    for input_dim, output_dim in [(1, 64), (64, 128)]:
+        layers += [
+            torch.nn.Conv2d(
+                input_dim, output_dim, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1)
+            ),
+            torch.nn.ReLU(inplace=True),
+            torch.nn.MaxPool2d(
+                kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False
+            ),
+        ]
+
+    for input_dim, output_dim in [(128, 256), (256, 512)]:
+        layers += [
+            torch.nn.Conv2d(
+                input_dim, output_dim, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1)
+            ),
+            torch.nn.ReLU(inplace=True),
+            torch.nn.Conv2d(
+                output_dim,
+                output_dim,
+                kernel_size=(3, 3),
+                stride=(1, 1),
+                padding=(1, 1),
+            ),
+            torch.nn.ReLU(inplace=True),
+            torch.nn.MaxPool2d(
+                kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False
+            ),
+        ]
+
+    return torch.nn.Sequential(*layers)
+
+
+def _build_embedding_network():
+    return torch.nn.Sequential(
+        torch.nn.Linear(512 * 4 * 6, 4096),
+        torch.nn.ReLU(True),
+        torch.nn.Linear(4096, 4096),
+        torch.nn.ReLU(True),
+        torch.nn.Linear(4096, 128),
+        torch.nn.ReLU(True),
+    )
+
+
+class VGGish(torch.nn.Module):
+    """Implementation of VGGish model :cite:`45611`."""
+    def __init__(self):
+        super().__init__()
+
+        self.features_network = _build_features_network()
+        self.embedding_network = _build_embedding_network()
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            input (torch.Tensor): batch of spectrograms, with shape `(n_example, 1, n_frame, 64)`.
+
+        Returns:
+            torch.Tensor: model output, with shape `(n_example, 128)`.
+        """
+        x = self.features_network(input)
+
+        x = x.permute(0, 2, 3, 1)
+        x = x.reshape(x.size(0), -1)
+
+        return self.embedding_network(x)
+
+
+class VGGishInputProcessor:
```

Comment:
This class is completely stateless. None of the method really require `self`. So I think it is better to extract them as plain functions.